<a href="https://colab.research.google.com/github/oreoluwaodeyinka/IntroToMachineLearning/blob/main/IntroMLAssignment1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Oreoluwa Odeyinka
# Intro To Machine Learning Assignment 1



*   Link to reading + code-along: https://www.datacamp.com/tutorial/essentials-linear-regression-python
*   Link to dataset: https://www.kaggle.com/c/boston-housing

## Assignment Outline
The goal of this assignment is to follow along with the code snippets above and create a python notebook that compares two versions of a linear regression model for the Boston housing dataset (you can copy and paste code for this).

*   Try and annotate each cell with a (really) brief explanation as to what that code chunk is doing in each cell.
*   After you have built two models, provide a brief description of how each of the models compare/contrast to one another.
*   Aside from feature selection, any other way that you can think of (or google) to compare two models would be fine as well.
*   After you are done, please upload your .ipynb python notebook file to your github account and submit the link for where your project lives here in google classroom for this assignment.





In [47]:
#Import necessary libraries

import pandas as pd
import numpy as np


In [48]:
#importing the statsmodels library
import statsmodels.api as sm
from sklearn import datasets

In [49]:
#Reading the CSV file
df = pd.read_csv('https://raw.githubusercontent.com/oreoluwaodeyinka/IntroToMachineLearning/refs/heads/main/train.csv')
df.head(5)

,ID,crim,zn,indus,chas,nox,rm,age,dis,rad,tax,ptratio,black,lstat,medv
0,1,0.00632,18.0,2.31,0,0.538,6.575,65.2,4.0900,1,296,15.3,396.90,4.98,24.0
1,2,0.02731,0.0,7.07,0,0.469,6.421,78.9,4.9671,2,242,17.8,396.90,9.14,21.6
2,4,0.03237,0.0,2.18,0,0.458,6.998,45.8,6.0622,3,222,18.7,394.63,2.94,33.4
3,5,0.06905,0.0,2.18,0,0.458,7.147,54.2,6.0622,3,222,18.7,396.90,5.33,36.2
4,7,0.08829,12.5,7.87,0,0.524,6.012,66.6,5.5605,5,311,15.2,395.60,12.43,22.9


In [50]:
# Set the features
# Exclude 'ID' and 'medv' (target) from features
feature_columns = [col for col in df.columns if col not in ['ID', 'medv']]
setDf = df[feature_columns]

# Set the target
target = df[["medv"]]

## First Model

This model is an OLS models so it is attempting to tell us how one input variable, which in this case is "rm" which is average number of rooms per dwelling connects to and helps predict "medv" which is the median value of owner-occupied homes in $1000's

In [51]:
#This is a section that attempts to create the first regression model
X = df["rm"]
y = target["medv"]

# Fit and make the predictions by the model
model = sm.OLS(y, X).fit()
predictions = model.predict(X)

# Print out the statistics
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                   medv   R-squared (uncentered):                   0.904
Model:                            OLS   Adj. R-squared (uncentered):              0.903
Method:                 Least Squares   F-statistic:                              3117.
Date:                Wed, 05 Aug 2026   Prob (F-statistic):                   8.25e-171
Time:                        04:05:29   Log-Likelihood:                         -1148.5
No. Observations:                 333   AIC:                                      2299.
Df Residuals:                     332   BIC:                                      2303.
Df Model:                           1                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
rm             3.7004      0.066     55.829      0.000       3.570       3.831
==============================================================================
Omnibus:                       66.008   Durbin-Watson:                   0.544
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              129.272
Skew:                           1.049   Prob(JB):                     8.49e-29
Kurtosis:                       5.216   Cond. No.                         1.00
==============================================================================

Notes:
[1] R² is computed without centering (uncentered) since the model does not contain a constant.
[2] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

This model summary tells us a lot of things, most of which I don't quite understand yet, but from the reading I can gather a couple of things


*   R-squared: 0.904. R-squared will give you an estimate of the relationship between movements of a dependent variable based on an independent variable's movements.
*   Coefficient: 3.7004. This means that if the RM variable increases by 1, the predicted value of medv increases by 3.7004.



#Second Model

This model is a lasso models so it adds an absolute penalty to the model that shrinks weak feature coefficients down to exactly zero which will in essence "mute the noise" by not taking them into account and help us get a better understanding of how well the input variable "rm" connects to and helps predict "medv".

In [52]:
from sklearn.linear_model import Lasso

# X contains your input feature (rm whch is average number of rooms)
# y contains your target values (medv which is median home value)

# alpha controls the penalty strength (higher alpha = more zeros)
lasso_model = Lasso(alpha=1.0)

# Ensure X is 2D for sklearn models
# If X is currently a pandas Series, convert it to a DataFrame (2D)
if isinstance(X, pd.Series):
    X = X.to_frame()

# Train the model
lasso_model.fit(X, y)

# Print R-squared score
print("Lasso R-squared:", lasso_model.score(X, y))

# Print coefficients
print("Lasso Coefficients:", lasso_model.coef_)




Lasso R-squared: 0.45142095228309875
Lasso Coefficients: [6.9623752]


**This is the explanation gemini gave me on Lasso regression**

*   Lasso R-squared: The R-squared value of 0.4514 means that approximately 45.14% of the variance in the 'medv' (median house value) can be explained by the 'rm' feature in this Lasso model.

*   The output [6.9623752] is the coefficient for the 'rm' (average number of rooms) feature. This value indicates the impact of the 'rm' feature on the 'medv' (median house value) according to the Lasso model, with a positive coefficient suggesting that as the number of rooms increases, the median house value also tends to increase. So if the  RM variable increases by 1, the predicted value of medv increases by 6.9623752.




**What that explanation leads me to believe**

Ok so the R-squared of the OLS model was about double that of the lasso model which means that the OLS model shows more variance.

#Model 3

Training the Lasso Model with all the varaibles not just "rm" so that I can actually see the model zeroing out what's not important.

In [53]:
from sklearn.linear_model import Lasso


# X contains your input feature (all of the variables this time minus medv)
# y contains your target values (medv which is median home value)

# alpha controls the penalty strength (higher alpha = more zeros)
lasso_model_all = Lasso(alpha=1.0)

# Ensure setDf is 2D for sklearn models
# If setDf is currently a pandas Series, convert it to a DataFrame (2D)
# Note: setDf should already be a DataFrame from earlier steps, but this ensures robustness.
if isinstance(setDf, pd.Series):
    setDf = setDf.to_frame()

# Train the model
lasso_model_all.fit(setDf, y)

# Print R-squared score, using the same features (setDf) as for fitting
print("Lasso R-squared:", lasso_model_all.score(setDf, y))

# Print coefficients
print("Lasso Coefficients:", lasso_model_all.coef_)

Lasso R-squared: 0.6699939563435349
Lasso Coefficients: [-0.          0.04939798  0.          0.         -0.          0.76582134
  0.02716634 -0.62762508  0.27197121 -0.01476439 -0.64923277  0.01141128
 -0.82752243]


Ok the reason I did this section was just to test out the lasso model to see what variables it would zero out.

These are all the variables it should have been looking at:
[crim,	zn,	indus,	chas,	nox,	rm,	age,	dis,rad, tax,	ptratio, black, lstat]

And these are the corresponding coefficients:
[-0.          0.04939798  0.          0.         -0.          0.76582134
  0.02716634 -0.62762508  0.27197121 -0.01476439 -0.64923277  0.01141128
 -0.82752243].

 It gave rm a coefficient of 0.76582134 which checks out because rm does have a high impact on medv as shown in the first model. So overall was just playing around with this section but it seems to have worked.



In [54]:
'''
##This section is for trying out some more of the examples from datacamp.
#trying out the linear model class

from sklearn import linear_model
X = df
y = target["medv"]

lm = linear_model.LinearRegression()
model = lm.fit(X,y)
predictions = lm.predict(X)
print(predictions[0:5])
'''


' \n##This section is for trying out some more of the examples from datacamp. \n#trying out the linear model class\n\nfrom sklearn import linear_model\nX = df\ny = target["medv"]\n\nlm = linear_model.LinearRegression()\nmodel = lm.fit(X,y)\npredictions = lm.predict(X)\nprint(predictions[0:5])\n'